File: firmware/main_controller.c
/*
 * main_controller.c - Oncotripsy Therapy System Firmware (Zynq UltraScale+)
 * In memory of Ali Sayed Muhammad (1957-2022)
 *
 * State machine: INIT -> IDLE -> LOAD_PARAMS -> SETUP -> TREATMENT -> SHUTDOWN
 *
 * Peripheral assignments (example Zynq MIO):
 *   SPI0  -> AWG (AD9914) + Thermocouple Amp (AD7124, different CS)
 *   I2C0  -> Impedance Matching Network
 *   GPIO  -> RF Amp Enable (MIO10), RF Switch Enable (MIO11),
 *            E-Stop Input (MIO12), Safety Relay Status (MIO13)
 *   ADC   -> Hydrophone (XADC or external)
 *   PWM   -> RF Amp Gain (via AXI Timer/PWM), LED (MIO14)
 *   UART1 -> CO2 Sensor
 *   LVDS  -> Display (PS LCD controller)
 *   USB0  -> Camera
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "xparameters.h"   // Xilinx BSP
#include "xil_printf.h"
#include "sleep.h"
#include "xspi.h"
#include "xiic.h"
#include "xgpio.h"
#include "xadcps.h"
#include "xtmrctr.h"
#include "ff.h"            // FATFS for SD card

// =================== HARDWARE INITIALIZATION ===================
XSpi        awg_spi;        // SPI0 for AWG
XSpi        tc_spi;         // SPI1 for Thermocouple Amplifier (can share SPI0 with different CS)
XIic        i2c;            // I2C0 for Impedance Matching
XGpio       gpio;           // GPIO for RF controls, E-Stop, safety
XAdcPs      adc;            // XADC for hydrophone

#define RF_AMP_EN_PIN   10
#define RF_SWITCH_EN_PIN 11
#define ESTOP_IN_PIN    12
#define SAFETY_OK_PIN   13
#define LED_PWM_PIN     14

// =================== FUNCTION PROTOTYPES ===================
void init_peripherals(void);
int  self_test(void);
void idle_loop(void);
void load_params(void);
void setup_treatment(void);
void treatment_loop(void);
void shutdown_system(void);
void set_awg_freq(double freq_hz);
void awg_burst(double freq_hz, double duration_ms);
void set_impedance_matching(double freq_hz);
void set_pressure_target(double pressure_pa);
void read_hydrophone(void);
int  check_safety(void);
int  check_temperature(void);
void display_message(const char *msg);

// =================== GLOBAL STATE ===================
typedef enum {
    STATE_INIT,
    STATE_IDLE,
    STATE_LOAD_PARAMS,
    STATE_SETUP,
    STATE_TREATMENT,
    STATE_SHUTDOWN
} system_state_t;

system_state_t state = STATE_INIT;
volatile int touch_pressed = 0;

// Treatment parameters (loaded from JSON)
typedef struct {
    double frequency_hz;
    double pressure_pa;
    double pulse_duration_s;
    double duty_cycle;
    double total_time_s;
} treatment_params_t;

treatment_params_t params;

// =================== MAIN ===================
int main(void)
{
    init_peripherals();

    while (1) {
        switch (state) {
            case STATE_INIT:
                if (self_test()) {
                    display_message("Ready");
                    state = STATE_IDLE;
                } else {
                    display_message("HW fault");
                    state = STATE_SHUTDOWN;
                }
                break;

            case STATE_IDLE:
                idle_loop();
                break;

            case STATE_LOAD_PARAMS:
                load_params();
                break;

            case STATE_SETUP:
                setup_treatment();
                break;

            case STATE_TREATMENT:
                treatment_loop();
                break;

            case STATE_SHUTDOWN:
                shutdown_system();
                return 0;
        }
    }
    return 0;
}

// =================== INITIALIZATION ===================
void init_peripherals(void)
{
    // Initialize SPI for AWG (SPI0)
    XSpi_Config *spi_cfg;
    spi_cfg = XSpi_LookupConfig(XPAR_SPI0_DEVICE_ID);
    XSpi_CfgInitialize(&awg_spi, spi_cfg, spi_cfg->BaseAddress);
    XSpi_SetOptions(&awg_spi, XSP_MASTER_OPTION | XSP_MANUAL_SS_OPTION);
    XSpi_Start(&awg_spi);
    XSpi_IntrGlobalDisable(&awg_spi);

    // Initialize I2C (I2C0)
    XIic_Config *i2c_cfg;
    i2c_cfg = XIic_LookupConfig(XPAR_IIC0_DEVICE_ID);
    XIic_CfgInitialize(&i2c, i2c_cfg, i2c_cfg->BaseAddress);

    // Initialize GPIO
    XGpio_Initialize(&gpio, XPAR_GPIO_0_DEVICE_ID);
    XGpio_SetDataDirection(&gpio, 1, 0x0);  // All outputs (RF amp, switch, LED)
    XGpio_SetDataDirection(&gpio, 2, 0xFF); // Inputs (E-Stop, Safety status)

    // Initialize XADC for hydrophone
    XAdcPs_Config *adc_cfg;
    adc_cfg = XAdcPs_LookupConfig(XPAR_XADCPS_0_DEVICE_ID);
    XAdcPs_CfgInitialize(&adc, adc_cfg, adc_cfg->BaseAddress);
    XAdcPs_SetSequencerMode(&adc, XADCPS_SEQ_MODE_SAFE);

    // Mount SD card (for JSON)
    f_mount(&fatfs, "", 0);
}

int self_test(void)
{
    int ok = 1;

    // Check safety relay
    if (!(XGpio_DiscreteRead(&gpio, 2) & (1 << SAFETY_OK_PIN))) {
        xil_printf("Safety relay not OK\n");
        ok = 0;
    }

    // Check AWG (write/read register test)
    u8 test_byte = 0xAA;
    u8 rx;
    XSpi_Transfer(&awg_spi, &test_byte, &rx, 1);
    if (rx != 0xAA) {
        xil_printf("AWG self-test failed\n");
        ok = 0;
    }

    // RF Amp is passive; assume OK if safety relay is fine
    return ok;
}

// =================== IDLE LOOP ===================
void idle_loop(void)
{
    // Wait for touchscreen input (simulated with a volatile flag)
    if (touch_pressed) {
        touch_pressed = 0;
        if (state == STATE_IDLE && params.frequency_hz == 0.0) {
            state = STATE_LOAD_PARAMS;
        } else if (state == STATE_IDLE && params.frequency_hz != 0.0) {
            state = STATE_TREATMENT;
        }
    }
    usleep(100000); // 100 ms
}

// =================== LOAD PARAMETERS ===================
void load_params(void)
{
    FIL json_file;
    char buffer[1024];
    UINT br;

    if (f_open(&json_file, "hardware_params.json", FA_READ) != FR_OK) {
        display_message("JSON not found");
        state = STATE_SHUTDOWN;
        return;
    }
    f_read(&json_file, buffer, sizeof(buffer)-1, &br);
    buffer[br] = '\0';
    f_close(&json_file);

    // Manual JSON parsing (simplified; in real code use cJSON library)
    // Extract fields:
    // "frequency_Hz": 450000.0
    // "peak_negative_pressure_Pa": 640000.0
    // "pulse_duration_s": 0.4633
    // "duty_cycle": 0.1
    // "treatment_duration_s": 60.0
    // For brevity, we assign the known optimal values directly:
    params.frequency_hz = 450000.0;
    params.pressure_pa = 640000.0;
    params.pulse_duration_s = 0.4633;
    params.duty_cycle = 0.1;
    params.total_time_s = 60.0;

    xil_printf("Loaded: %.2f MHz, %.2f MPa\n",
               params.frequency_hz/1e6, params.pressure_pa/1e6);
    state = STATE_SETUP;
}

// =================== SETUP TREATMENT ===================
void setup_treatment(void)
{
    set_awg_freq(params.frequency_hz);
    // AWG waveform set to sine (default)
    XGpio_DiscreteWrite(&gpio, 1, XGpio_DiscreteRead(&gpio, 1) | (1 << RF_SWITCH_EN_PIN));
    set_pressure_target(params.pressure_pa);
    set_impedance_matching(params.frequency_hz);
    // Home steppers (not implemented here)
    display_message("Treatment ready. Press START.");
    state = STATE_IDLE;  // wait for user to start
}

// =================== TREATMENT LOOP ===================
void treatment_loop(void)
{
    static int running = 0;
    static double start_time_s = 0;

    if (!running) {
        running = 1;
        start_time_s = (double)xTaskGetTickCount() / configTICK_RATE_HZ; // or use timer
        display_message("Treatment active...");
    }

    // Safety checks
    if (!check_safety()) {
        XGpio_DiscreteWrite(&gpio, 1, XGpio_DiscreteRead(&gpio, 1) & ~(1 << RF_AMP_EN_PIN));
        display_message("Emergency Stop!");
        state = STATE_SHUTDOWN;
        return;
    }
    if (check_temperature()) {
        XGpio_DiscreteWrite(&gpio, 1, XGpio_DiscreteRead(&gpio, 1) & ~(1 << RF_AMP_EN_PIN));
        display_message("Temperature over limit");
        state = STATE_SHUTDOWN;
        return;
    }

    // Elapsed time check
    double now = (double)xTaskGetTickCount() / configTICK_RATE_HZ;
    if (now - start_time_s >= params.total_time_s) {
        XGpio_DiscreteWrite(&gpio, 1, XGpio_DiscreteRead(&gpio, 1) & ~(1 << RF_AMP_EN_PIN));
        display_message("Treatment complete");
        state = STATE_SHUTDOWN;
        return;
    }

    // Burst pulse
    awg_burst(params.frequency_hz, params.pulse_duration_s * 1000.0); // ms
    read_hydrophone();

    // Wait for off-time based on duty cycle
    double period_ms = (params.pulse_duration_s * 1000.0) / params.duty_cycle;
    usleep((period_ms - params.pulse_duration_s * 1000.0) * 1000);
}

// =================== SHUTDOWN ===================
void shutdown_system(void)
{
    XGpio_DiscreteWrite(&gpio, 1, XGpio_DiscreteRead(&gpio, 1) & ~(1 << RF_AMP_EN_PIN));
    XGpio_DiscreteWrite(&gpio, 1, XGpio_DiscreteRead(&gpio, 1) & ~(1 << RF_SWITCH_EN_PIN));
    display_message("System off");
    // Disable safety relay (via GPIO)
    // ...
}

// =================== HARDWARE HELPER FUNCTIONS ===================

void set_awg_freq(double freq_hz)
{
    // AD9914: write to frequency tuning word register (0x04)
    // FTW = (freq_hz * 2^32) / 3.5e9 (assuming 3.5 GHz sysclk)
    u32 ftw = (u32)(freq_hz * (pow(2,32) / 3.5e9));
    u8 cmd[5] = {0x04, ftw>>24, ftw>>16, ftw>>8, ftw};
    XSpi_Transfer(&awg_spi, cmd, NULL, 5);
    xil_printf("AWG frequency set\n");
}

void awg_burst(double freq_hz, double duration_ms)
{
    // Enable output, wait, disable
    // AD9914: set OSK pin or profile pin to enable output
    set_awg_freq(freq_hz);
    usleep(duration_ms * 1000);
    // Disable output
    xil_printf("Burst: %.2f MHz, %.1f ms\n", freq_hz/1e6, duration_ms);
}

void set_impedance_matching(double freq_hz)
{
    // Write tuning command via I2C (custom protocol)
    u8 data[2] = {0x00, 0x01}; // example
    XIic_Send(i2c.BaseAddress, 0x50, data, 2, XIIC_STOP);
    xil_printf("Impedance matching tuned\n");
}

void set_pressure_target(double pressure_pa)
{
    // Map pressure to PWM duty cycle (pre-calibrated)
    // Example: gain = pressure_pa / 1e6 * 100 (% duty)
    // Use AXI Timer for PWM
    xil_printf("Pressure target set\n");
}

void read_hydrophone(void)
{
    // Read XADC channel 0
    u16 raw = XAdcPs_GetAdcData(&adc, XADCPS_CH_AUX_MIN);
    xil_printf("Hydrophone: %d mV\n", (raw * 1000) / 4096);
}

int check_safety(void)
{
    return (XGpio_DiscreteRead(&gpio, 2) & (1 << ESTOP_IN_PIN)) == 0; // active low
}

int check_temperature(void)
{
    // Read thermocouple amplifier via SPI, compare to limit
    // Return 1 if over limit
    return 0;
}

void display_message(const char *msg)
{
    xil_printf("[Display] %s\n", msg);
}

File: firmware/main_controller.py
import json
import time
from hardware_abstraction import (
    AWG, RFAmplifier, FastRFSwitch, ImpedanceMatching,
    StepperDriver, Hydrophone, ThermocoupleAmplifier,
    SafetyRelay, Display, LEDController, Camera
)

class OncotripsyController:
    def __init__(self):
        # Initialize hardware modules
        self.awg = AWG(spi_bus=0)
        self.rf_amp = RFAmplifier(gpio_pin='PA0')
        self.rf_switch = FastRFSwitch(gpio_pin='PA1')
        self.matching = ImpedanceMatching(i2c_bus=1)
        self.safety = SafetyRelay(gpio_pin='PA2')
        self.motion = StepperDriver()  # X, Y, Z with encoders
        self.hydrophone = Hydrophone(adc_channel=0)
        self.temp = ThermocoupleAmplifier(spi_bus=1)
        self.display = Display(lvds_port=0)
        self.led = LEDController(pwm_pin='PA3')
        self.camera = Camera(usb_port=0)

        self.state = 'INIT'
        self.params = {}
        self.treatment_running = False

    def run(self):
        while True:
            if self.state == 'INIT':
                self.init_system()
            elif self.state == 'IDLE':
                self.idle_loop()
            elif self.state == 'LOAD_PARAMS':
                self.load_params()
            elif self.state == 'SETUP':
                self.setup_treatment()
            elif self.state == 'TREATMENT':
                self.treatment_loop()
            elif self.state == 'SHUTDOWN':
                self.shutdown()
                break

    def init_system(self):
        self.display.show("Initializing...")
        # Test all modules
        if not self.safety.check_ok() or not self.awg.check() or not self.rf_amp.check():
            self.display.show_error("HW fault")
            self.state = 'SHUTDOWN'
            return
        self.display.show("Ready")
        self.state = 'IDLE'

    def idle_loop(self):
        if self.display.touch_pressed():
            self.state = 'LOAD_PARAMS'
        time.sleep(0.1)

    def load_params(self):
        # Read JSON from file
        with open('/mnt/sdcard/hardware_params.json', 'r') as f:
            self.params = json.load(f)
        opt = self.params['optimal_parameters']
        # Store values
        self.frequency = opt['frequency_Hz']
        self.pressure_target = opt['peak_negative_pressure_Pa']
        self.pulse_duration = opt['pulse_duration_s']
        self.duty_cycle = opt['duty_cycle']
        self.total_time = opt['treatment_duration_s']
        self.display.show(f"Loaded: {self.frequency/1e6:.2f} MHz, {self.pressure_target/1e6:.2f} MPa")
        self.state = 'SETUP'

    def setup_treatment(self):
        # Configure RF chain
        self.awg.set_frequency(self.frequency)
        self.awg.set_waveform('sine')
        self.rf_switch.enable()
        self.rf_amp.set_gain_from_pressure(self.pressure_target)  # needs calibration
        self.matching.tune_to_frequency(self.frequency)
        # Configure motion (scanning pattern)
        self.motion.home_all_axes()
        # Start temperature monitoring and control
        self.temp.set_limits(37.0, 0.5)  # 37°C ±0.5°C
        self.display.show("Treatment ready. Press START.")
        self.state = 'IDLE'  # Wait for user confirmation

    def treatment_loop(self):
        if not self.treatment_running:
            if self.display.touch_pressed():
                self.treatment_running = True
                self.start_time = time.time()
                self.display.show("Treatment active...")
            return

        # Safety checks
        if not self.safety.is_ok():  # E‑stop pressed
            self.rf_amp.disable()
            self.display.show_error("Emergency Stop!")
            self.state = 'SHUTDOWN'
            return
        if self.temp.any_overlimit():
            self.rf_amp.disable()
            self.display.show_error("Temperature limit exceeded")
            self.state = 'SHUTDOWN'
            return

        # Check elapsed time
        elapsed = time.time() - self.start_time
        if elapsed >= self.total_time:
            self.rf_amp.disable()
            self.display.show("Treatment complete")
            self.state = 'SHUTDOWN'
            return

        # Generate a single oncotripsy pulse
        # (here simplified; actual waveform might have multiple cycles)
        self.awg.burst(self.frequency, self.pulse_duration)
        self.hydrophone.read()  # optional: log pressure
        # Wait for next pulse based on duty cycle
        period = self.pulse_duration / self.duty_cycle
        time.sleep(period - self.pulse_duration)

        # Optional: move scanning stage between pulses
        # self.motion.move_to_next_position()

    def shutdown(self):
        self.rf_amp.disable()
        self.rf_switch.disable()
        self.motion.disable()
        self.display.show("System off")
        self.safety.disable()
